pieces/rest_api_postgres/README.md

```diff
@@ -0,0 +1,135 @@
+restart the server any time code is changed. This prevents us from having to stop and start the
+server any time a code change is made. If you do not already have `cargo-watch` installed you can
+install it by running `cargo install cargo-watch`.
+
+Install dependencies via Cargo. In this piece, we use the `anyhow`, `axum`, `chrono`, `dotenvy`,
+`serde`, `sqlx`, and `tokio` crates. Addtinally we use `http-body-util`, `mime`, `serde_json`,
+`tower`, and `uuid` for testing.
+
+```sh
+cargo add anyhow
+cargo add axum
+cargo add chrono --features chrono
+cargo add dotenvy
+cargo add serde --features derive
+cargo add sqlx --features postgres --features runtime-tokio --features chrono
+cargo add tokio --features full
+cargo add --dev http-body-util
+cargo add --dev mime
+cargo add --dev serde-json
+cargo add --dev tower
+cargo add --dev uuid --features v4
+```
+
+# Start the database
+
+From the `pieces/rest_api_postgres` directory start the database with docker compose
+
+```sh
+docker compose up
+```
+
+## Start the development server
+
+The server can be started either with `Make` or manually.
+
+```sh
+# With Make
+make dev
+
+# Manually
+cargo watch -w src -c -x run
+```
+
+Make commands are provided to interact with the API
+
+```sh
+# Add a book
+make add-book BOOK='{"title": "Some Title", "authorFirstName": "FirstName", "authorLastName": "LastName", "bookStatus": "Read", "dateAdded": "2024-01-30T15:47:01Z", "dateRead": null, "rating": 1}'
+
+# Delete a book
+make delete-book BOOK_ID=1
+
+# Get a book
+make get-book BOOK_ID=1
+
+# Get all books
+make get-books
+
+# Update a book
+make update-book BOOK='{"id": 2, "title": "New Title", "authorFirstName": "FirstName", "authorLastName": "LastName", "bookStatus": "Read", "dateAdded": "2024-01-30T15:47:01Z", "dateRead": null, "rating": 1}'
+```
+
+# Takeaways
+
+With FastAPI we are using `uvicorn` the `--reload` flag to auto-restart the server when changes are
+made. In Rust we get the equivalent functionality by using `cargo-watch`. While in this example we are
+using `cargo-watch` to restart the webserver, `cargo-watch` is not specific to web servers and can
+be used for any Rust program that you want to re-run when changes are saved.
+
+One nice feature of FastAPI is OpenAPI documenation is automatically generated for routes. This
+makes it very easy to test and document the API. Because Axum does not provide this same functionality
+a program such as [postman](https://www.postman.com/) with test routes manually created is needed to
+serve this pourpose. Additionally documentation would still need to be created.
+
+In the tests you may have noticed that in Python we clear the database tables between each test, but
+we don't do this in Rust. The reason for this is by default pytest runs one test at a time while
+the default in Rust is to run tests in parallel. Because of this if we were to clear the tables
```

Comment:
Rust tests are equivalent to running pytest with the xdist plugin. So they run across cores even if the code isn't async. You would still call this asynchronous rather than parallel?

Weird that you are getting that error with tests, it should be impossible. The titles are set with uuids so they should never be the same. I'll double check and make sure I didn't miss one somewhere.

Reply:
You're right, that was my misconception. The [book](https://doc.rust-lang.org/book/ch11-02-running-tests.html#running-tests-in-parallel-or-consecutively) says that the tests run in parallel (no mention of async). I'll remove those suggestions.